# 第7回　ワークブック

Python II：CSV、pandas、plot

2026年10月27日（火）3限　金融論

## 使い方

上から順に、1つずつ実行していくnotebookです。

- 実行する　code cellを選び、Shift+Enterで実行します。押す前に、何が出るか予想します。
- やってみる　すぐ下の `# ここに書きます` のcellに自分で書いて実行します。既にあるcellを書き換える指示もあります。

errorを確かめるcellは、確かめたら行頭に `#` を付けておきます。8節で全部を実行し直すとき、errorで止まらなくなります。

詰まったら、cellの内容とエラーの表示をCodexに見せて聞けます。「やってみる」の答え合わせも頼めます。

使うデータは `data/` の2つのCSVです。どちらも架空の3社の株価で、実在の企業とは関係ありません。4節と6節で `output/` folderができ、書き出したCSVが入ります。

`workbook.py` には、「実行する」のcodeが上から順に入っています。terminalで `python workbook.py` と打つと、notebookと同じ処理を単独で実行できます。

## 準備

1. `workbook/` folderを、自分の作業場所にコピーします。`data/` も一緒にコピーされます。
2. VS Codeでfolderごと開きます（File → Open Folder）。
3. このnotebookを開き、右上の Select Kernel → Python Environments から `finance-env` を選びます。
4. 下のcellを実行します。pathに `finance-env` が含まれ、pandasとmatplotlibのversionが表示されれば準備完了です。

`No module named 'pandas'` と出たら、選んだkernelにpandasが入っていません。入れる手順はCodexに頼めます。

In [1]:
import sys
print(sys.executable)

import pandas as pd
import matplotlib
print(pd.__version__, matplotlib.__version__)

[検証に使用したPython]


2.2.3 3.10.8


# 1. 第6回の復習

## 1-1　listとfor

架空の3社の銘柄コードをlistに入れ、1つずつ表示します。

実行する

In [2]:
codes = ["01110", "02220", "03330"]
for code in codes:
    print(code, len(code))

01110 5
02220 5
03330 5


銘柄コードは5文字の文字列です。先頭の0も1文字に数えます。

やってみる

1. `print(codes[0])` を実行すると `01110` が出ます。
2. `print(int(codes[0]))` を実行すると `1110` になります。数に変換すると、先頭の0は消えます。3節で、CSVを読むときに同じことが起きます。

In [3]:
# ここに書きます

## 1-2　関数とreturn

前の日の終値 `old` から今日の終値 `new` への変化率（%）を返す関数です。

実行する

In [4]:
def pct_change(old, new):
    return (new - old) / old * 100

change = pct_change(2600, 2670)
print(round(change, 2))

2.69


2,600円から2,670円への変化率は2.69%です。`return` で返した値を `change` に入れてから丸めています。

やってみる

引数の順番を入れ替えて `pct_change(2670, 2600)` を計算します。-2.62になり、符号だけを変えた-2.69にはなりません。分母が2,670円に変わるからです。

In [5]:
# ここに書きます

# 2. CSVと観測単位

## 2-1　CSVをtextとして見る

`data/prices_short.csv` は、架空の3社の3日分の終値と出来高です。

実行する

In [6]:
print(open("data/prices_short.csv", encoding="utf-8").read())

Date,Code,C,Vo
2024/8/5,01110,"2,467",27465800
2024/8/5,02220,"2,391",10278200
2024/8/5,03330,,0
2024/8/2,01110,"2,492",28425400
2024/8/2,02220,"2,322",10757000
2024/8/2,03330,"3,914",1936800
2024/8/1,01110,"2,454",20507300
2024/8/1,02220,"2,364",11575200
2024/8/1,03330,"3,872",5279400



1行目は列の名前（見出し）です。2行目から下が値で、`,` で区切られています。`"2,467"` が引用符で囲まれているのは、値の中の `,` を区切りと区別するためです。`03330` の2024年8月5日は、終値 `C` が空です。

見出しを除く9行は、3社×3日です。1行が「1つの銘柄の1日」を表しています。

やってみる

VS CodeのExplorerで `data/prices_short.csv` をクリックすると、上と同じ文字がそのまま表示されます。CSVはtext fileなので、editorで中身を確かめられます。

## 2-2　Excelで開く

やってみる

`data/prices_short.csv` をExcelで開きます。

- `01110` が `1110` になります。Excelが数だと判断して、先頭の0を落とすからです。
- `2024-08-05` は日付として扱われます。表示の形はPCの設定によって変わります。
- `"2,467"` は数の2467になり、セルの右に寄ります。

Excelで上書き保存すると、この形でCSVが書き換わります。確かめたら、保存せずに閉じます。

# 3. pandasで読む

## 3-1　read_csv

実行する

In [7]:
import pandas as pd

short = pd.read_csv("data/prices_short.csv")
short

,Date,Code,C,Vo
0,2024/8/5,1110,"2,467",27465800
1,2024/8/5,2220,"2,391",10278200
2,2024/8/5,3330,NaN,0
3,2024/8/2,1110,"2,492",28425400
4,2024/8/2,2220,"2,322",10757000
5,2024/8/2,3330,"3,914",1936800
6,2024/8/1,1110,"2,454",20507300
7,2024/8/1,2220,"2,364",11575200
8,2024/8/1,3330,"3,872",5279400


`import pandas as pd` で、pandasというlibraryを `pd` という短い名前で使えるようにしています。`pd.read_csv` がCSV fileを読んで表（DataFrame）を作り、`short` に入れます。

CSV fileはdiskに保存された文字の並びで、DataFrameはkernelのメモリにある表です。`short` を書き換えても、CSV fileは変わりません。

やってみる

`short.head(3)` と `short.tail(2)` を実行します。先頭の3行と、最後の2行が出ます。

In [8]:
# ここに書きます

## 3-2　shapeとdtypes

実行する

In [9]:
print(short.shape)
print(short.dtypes)

(9, 4)
Date    object
Code     int64
C       object
Vo       int64
dtype: object


`(9, 4)` は9行4列です。`dtypes` は列ごとの型です。`Code` は `int64`（整数）で、表の中では `1110` になっています。`C` は `object` で、数ではなく文字列として読まれています。`"2,467"` に `,` が入っているからです。

やってみる

`short["Code"]` を実行します。`1110`、`2220`、`3330` が並び、先頭の0が消えています。

In [10]:
# ここに書きます

## 3-3　列を選ぶ

実行する

In [11]:
short["C"]

0    2,467
1    2,391
2      NaN
3    2,492
4    2,322
5    3,914
6    2,454
7    2,364
8    3,872
Name: C, dtype: object

In [12]:
short[["Date", "Code", "C"]]

,Date,Code,C
0,2024/8/5,1110,"2,467"
1,2024/8/5,2220,"2,391"
2,2024/8/5,3330,NaN
3,2024/8/2,1110,"2,492"
4,2024/8/2,2220,"2,322"
5,2024/8/2,3330,"3,914"
6,2024/8/1,1110,"2,454"
7,2024/8/1,2220,"2,364"
8,2024/8/1,3330,"3,872"


列の名前を1つ `[ ]` に入れると、その列だけが出ます。複数の列は、名前のlistを `[ ]` に入れます。外側の `[ ]` が選ぶ操作で、内側の `[ ]` が名前のlistです。

やってみる

1. `short[["Code", "Vo"]]` で、銘柄コードと出来高の2列を選びます。
2. `short["Close"]` を実行します。`Close` という列はないので `KeyError: 'Close'` になります。確かめたら行頭に `#` を付けます。

In [13]:
# ここに書きます

## 3-4　行を選ぶ

`Code` が `"01110"` の行だけを取り出します。

実行する

In [14]:
short[short["Code"] == "01110"]

,Date,Code,C,Vo


行が1つも出ません（Empty DataFrame）。`Code` 列は数の `1110` なので、文字列の `"01110"` と一致する行がないからです。

やってみる

1. `short["Code"] == "01110"` だけを実行します。9行すべてが `False` です。`short[ ]` の中にこの `True`・`False` の列を入れると、`True` の行だけが残ります。
2. `short[short["Code"] == 1110]` を実行します。数の1110と比べると、3行が出ます。

In [15]:
# ここに書きます

## 3-5　listとDataFrame、関数とmethod

実行する

In [16]:
vo_list = [27465800, 10278200, 0]
print(vo_list * 2)
print(short["Vo"] * 2)

[27465800, 10278200, 0, 27465800, 10278200, 0]
0    54931600
1    20556400
2           0
3    56850800
4    21514000
5     3873600
6    41014600
7    23150400
8    10558800
Name: Vo, dtype: int64


listに2を掛けると、listが2回繰り返されます。DataFrameの列に2を掛けると、行ごとに2倍になります。

実行する

In [17]:
print(len(short))
print(short.head(2))

9
       Date  Code      C        Vo
0  2024/8/5  1110  2,467  27465800
1  2024/8/5  2220  2,391  10278200


`len(short)` は、関数 `len` に `short` を渡しています。`short.head(2)` は、DataFrameが持っている `head` というmethodを呼んでいます。`変数名.method名(引数)` の形です。`pd.read_csv(...)` は、pandasの中の関数 `read_csv` を呼ぶ書き方です。

`short` はPythonの変数の名前で、`"Vo"` は表の中の列の名前です。列の名前は文字列なので、引用符で囲みます。

やってみる

`short.shape` と `short.shape()` を実行します。`shape` は値を持っているだけでmethodではないので、`( )` を付けると `TypeError: 'tuple' object is not callable` になります。確かめたら行頭に `#` を付けます。

In [18]:
# ここに書きます

# 4. 型・日付・欠損・並べ替え

## 4-1　銘柄コードを文字列で読む

実行する

In [19]:
short = pd.read_csv("data/prices_short.csv", dtype={"Code": str})
print(short.dtypes)
short[short["Code"] == "01110"]

Date    object
Code    object
C       object
Vo       int64
dtype: object


,Date,Code,C,Vo
0,2024/8/5,01110,"2,467",27465800
3,2024/8/2,01110,"2,492",28425400
6,2024/8/1,01110,"2,454",20507300


`dtype={"Code": str}` を付けて読むと、`Code` 列が文字列（`object`）になり、先頭の0が残ります。3-4で何も出なかった選び方で、3行が出ます。

やってみる

`short["Code"].str.len()` を実行します。すべて5です。`.str` を付けると、列の文字列1つ1つに文字列の操作ができます。

In [20]:
# ここに書きます

## 4-2　数に見える文字列を数にする

実行する

In [21]:
short["C"] * 2

0    2,4672,467
1    2,3912,391
2           NaN
3    2,4922,492
4    2,3222,322
5    3,9143,914
6    2,4542,454
7    2,3642,364
8    3,8723,872
Name: C, dtype: object

`"2,467"` が `2,4672,467` になります。文字列に2を掛けたので、文字が2回つながりました。空だった値は `NaN` のままです。

実行する

In [22]:
c_text = short["C"].str.replace(",", "")
short["C"] = pd.to_numeric(c_text)
print(short.dtypes)
short

Date     object
Code     object
C       float64
Vo        int64
dtype: object


,Date,Code,C,Vo
0,2024/8/5,01110,2467.0,27465800
1,2024/8/5,02220,2391.0,10278200
2,2024/8/5,03330,NaN,0
3,2024/8/2,01110,2492.0,28425400
4,2024/8/2,02220,2322.0,10757000
5,2024/8/2,03330,3914.0,1936800
6,2024/8/1,01110,2454.0,20507300
7,2024/8/1,02220,2364.0,11575200
8,2024/8/1,03330,3872.0,5279400


`str.replace(",", "")` で `,` を消し、`pd.to_numeric` で数に変換しています。`C` は `float64` になり、空だった値は `NaN` です。

やってみる

1. `short["C"] * 2` をもう一度実行します。`4934.0` のように、数として2倍になります。
2. `pd.to_numeric(pd.read_csv("data/prices_short.csv")["C"])` を実行します。`,` を消していないので `ValueError: Unable to parse string "2,467" at position 0` になります。確かめたら行頭に `#` を付けます。

In [23]:
# ここに書きます

## 4-3　日付を日付型にする

実行する

In [24]:
short["Date"] = pd.to_datetime(short["Date"])
print(short.dtypes)
short

Date    datetime64[ns]
Code            object
C              float64
Vo               int64
dtype: object


,Date,Code,C,Vo
0,2024-08-05,01110,2467.0,27465800
1,2024-08-05,02220,2391.0,10278200
2,2024-08-05,03330,NaN,0
3,2024-08-02,01110,2492.0,28425400
4,2024-08-02,02220,2322.0,10757000
5,2024-08-02,03330,3914.0,1936800
6,2024-08-01,01110,2454.0,20507300
7,2024-08-01,02220,2364.0,11575200
8,2024-08-01,03330,3872.0,5279400


`Date` が `datetime64[ns]` になります。表示は `2024-08-05` のまま変わりませんが、型が文字列から日付に変わっています。

やってみる

`sorted(["2024/8/5", "2024/10/1", "2024/9/30"])` を実行します。`['2024/10/1', '2024/8/5', '2024/9/30']` の順になり、10月が8月より前に来ます。文字列は左から1文字ずつ比べるので、`"1"` が `"8"` より先になるからです。日付型にすると、日付の順に並びます。

In [25]:
# ここに書きます

## 4-4　並べ替える

CSVは新しい日付が上にあります。銘柄ごとに、古い日付から並べ直します。

実行する

In [26]:
short = short.sort_values(["Code", "Date"]).reset_index(drop=True)
short

,Date,Code,C,Vo
0,2024-08-01,01110,2454.0,20507300
1,2024-08-02,01110,2492.0,28425400
2,2024-08-05,01110,2467.0,27465800
3,2024-08-01,02220,2364.0,11575200
4,2024-08-02,02220,2322.0,10757000
5,2024-08-05,02220,2391.0,10278200
6,2024-08-01,03330,3872.0,5279400
7,2024-08-02,03330,3914.0,1936800
8,2024-08-05,03330,NaN,0


`sort_values(["Code", "Date"])` は、まず `Code` で並べ、同じ `Code` の中では `Date` で並べます。`reset_index(drop=True)` は、左端の行番号を0から振り直します。

やってみる

`short.sort_values("Date", ascending=False)` を実行します。日付の新しい順になり、同じ日の3社が並びます。

In [27]:
# ここに書きます

## 4-5　欠損と0

実行する

In [28]:
print(short.isna().sum())
short[short["Code"] == "03330"]

Date    0
Code    0
C       1
Vo      0
dtype: int64


,Date,Code,C,Vo
6,2024-08-01,03330,3872.0,5279400
7,2024-08-02,03330,3914.0,1936800
8,2024-08-05,03330,NaN,0


`isna()` は値が欠損なら `True` を返し、`sum()` で列ごとに数えています。欠損は `C` の1つです。

Retail C（`03330`）の2024年8月5日は、終値が `NaN`、出来高が `0` です。この日は売買が成立しませんでした。終値は「値がない」ので欠損、出来高は「売買された株数が0株」なので0です。

実行する

In [29]:
retail = short[short["Code"] == "03330"]
print(retail["C"].mean())
print(retail["C"].fillna(0).mean())

3893.0
2595.3333333333335


`mean()` は欠損を飛ばし、残りの2日で平均します。3,893.0円です。`fillna(0)` で欠損を0円にしてから平均すると、3日で割って約2,595.33円になります。売買のなかった日の株価を0円として計算したことになります。

やってみる

`retail["Vo"].mean()` を実行します。2,405,400.0です。出来高の0は実際に0株だったので、平均に含まれます。

In [30]:
# ここに書きます

## 4-6　保存して読み直す

実行する

In [31]:
import os

os.makedirs("output", exist_ok=True)
short.to_csv("output/prices_short_clean.csv", index=False)

back = pd.read_csv("output/prices_short_clean.csv")
print(back.dtypes)
back.head(3)

Date     object
Code      int64
C       float64
Vo        int64
dtype: object


,Date,Code,C,Vo
0,2024-08-01,1110,2454.0,20507300
1,2024-08-02,1110,2492.0,28425400
2,2024-08-05,1110,2467.0,27465800


`to_csv` でDataFrameをCSV fileに書き出しています。`os.makedirs` は `output` folderを作り、すでにあれば何もしません。`index=False` は、左端の行番号を書き出さない指定です。

読み直すと、`Code` がまた `int64` になり、`Date` は `object` です。CSVに残るのは文字だけで、「この列は文字列」「この列は日付」という型の情報は保存されません。読むたびに `dtype` や日付の変換を付けると、同じ型に戻ります。

やってみる

VS Codeで `output/prices_short_clean.csv` を開くと、`01110` や `2024-08-01` が文字のまま入っています。`pd.read_csv("output/prices_short_clean.csv", dtype={"Code": str}, parse_dates=["Date"])` で読み直すと、`Code` は `object`、`Date` は `datetime64[ns]` になります。

In [32]:
# ここに書きます

# 5. 2年分の表を読む

## 5-1　dtypeとparse_datesを付けて読む

`data/prices.csv` は、架空の3社の2024年1月4日から2025年12月30日までの日次株価です。

実行する

In [33]:
prices = pd.read_csv("data/prices.csv", dtype={"Code": str}, parse_dates=["Date"])
print(prices.shape)
print(prices.dtypes)
prices.head()

(1545, 14)
Date         datetime64[ns]
Code                 object
O                   float64
H                   float64
L                   float64
C                   float64
Vo                  float64
Va                  float64
AdjFactor           float64
AdjO                float64
AdjH                float64
AdjL                float64
AdjC                float64
AdjVo               float64
dtype: object


,Date,Code,O,H,L,C,Vo,Va,AdjFactor,AdjO,AdjH,AdjL,AdjC,AdjVo
0,2024-01-04,01110,2604.0,2606.0,2598.0,2600.0,24659800.0,6.414836e+10,1.0,2604.0,2606.0,2598.0,2600.0,24659800.0
1,2024-01-05,01110,2599.0,2672.0,2597.0,2670.0,37151900.0,9.831631e+10,1.0,2599.0,2672.0,2597.0,2670.0,37151900.0
2,2024-01-08,01110,2670.0,2721.0,2669.0,2720.0,32006400.0,8.652397e+10,1.0,2670.0,2721.0,2669.0,2720.0,32006400.0
3,2024-01-09,01110,2716.0,2716.0,2679.0,2679.0,34073900.0,9.170422e+10,1.0,2716.0,2716.0,2679.0,2679.0,34073900.0
4,2024-01-10,01110,2680.0,2684.0,2617.0,2621.0,14109600.0,3.725875e+10,1.0,2680.0,2684.0,2617.0,2621.0,14109600.0


1,464行14列です。`parse_dates=["Date"]` を付けると、読むときに `Date` が日付型になります。

列の名前は、第8回で使うJ-Quants APIの株価と同じです。`O`・`H`・`L`・`C` は始値・高値・安値・終値（円）、`Vo` は出来高（株）、`Va` は売買代金（円）です。`AdjFactor` は株式分割などの調整係数で、`Adj` で始まる列は分割などを調整した値です。

やってみる

`prices.tail()` で最後の5行を見ます。最後の行は `03330` の2025-12-30で、終値は3,869.0円です。

In [34]:
# ここに書きます

## 5-2　銘柄ごとの行数

実行する

In [35]:
prices["Code"].value_counts()

Code
01110    515
02220    515
03330    515
Name: count, dtype: int64

3社とも488行です。1,464 = 3 × 488 なので、どの銘柄にも同じ日数の行があります。

やってみる

`prices["Date"].min()`、`prices["Date"].max()`、`prices["Date"].nunique()` を実行します。2024-01-04、2025-12-30、488です。

In [36]:
# ここに書きます

## 5-3　期間で絞る

実行する

In [37]:
aug = prices[(prices["Date"] >= "2024-08-01") & (prices["Date"] <= "2024-08-31")]
print(aug.shape)
aug.head()

(66, 14)


,Date,Code,O,H,L,C,Vo,Va,AdjFactor,AdjO,AdjH,AdjL,AdjC,AdjVo
150,2024-08-01,01110,2510.0,2512.0,2452.0,2454.0,20507300.0,5.070772e+10,1.0,2510.0,2512.0,2452.0,2454.0,20507300.0
151,2024-08-02,01110,2451.0,2499.0,2444.0,2492.0,28425400.0,7.044762e+10,1.0,2451.0,2499.0,2444.0,2492.0,28425400.0
152,2024-08-05,01110,2493.0,2496.0,2464.0,2467.0,27465800.0,6.799617e+10,1.0,2493.0,2496.0,2464.0,2467.0,27465800.0
153,2024-08-06,01110,2464.0,2518.0,2456.0,2510.0,35626700.0,8.887674e+10,1.0,2464.0,2518.0,2456.0,2510.0,35626700.0
154,2024-08-07,01110,2508.0,2537.0,2504.0,2533.0,15785200.0,3.985237e+10,1.0,2508.0,2537.0,2504.0,2533.0,15785200.0


`&` は「かつ」です。2つの条件を、それぞれ `( )` で囲みます。2024年8月は21営業日なので、3社で63行です。

やってみる

期間を2025年の1年間（`"2025-01-01"` から `"2025-12-31"`）に変えた表を `y2025` に入れます。729行になります。

In [38]:
# ここに書きます

## 5-4　列を足す

2024年度の教材の `mutate` と同じく、高値と安値の平均の列を作ります。

実行する

In [39]:
prices["HLmean"] = (prices["H"] + prices["L"]) / 2
prices[["Date", "Code", "H", "L", "HLmean"]].head(3)

,Date,Code,H,L,HLmean
0,2024-01-04,01110,2606.0,2598.0,2602.0
1,2024-01-05,01110,2672.0,2597.0,2634.5
2,2024-01-08,01110,2721.0,2669.0,2695.0


右辺の計算は行ごとに行われ、新しい列 `HLmean` に入ります。2024-01-04の `01110` は、(2,606 + 2,598) / 2 = 2,602.0です。

やってみる

高値と安値の差の列を `prices["HLrange"] = prices["H"] - prices["L"]` で作ります。2024-01-04の `01110` は8.0です。

In [40]:
# ここに書きます

## 5-5　銘柄コードに名前を付ける

実行する

In [41]:
names = {"01110": "Auto A", "02220": "Auto B", "03330": "Retail C"}
prices["Name"] = prices["Code"].map(names)
prices[["Date", "Code", "Name", "C"]].tail(3)

,Date,Code,Name,C
1542,2025-12-26,03330,Retail C,3880.0
1543,2025-12-29,03330,Retail C,3915.0
1544,2025-12-30,03330,Retail C,3869.0


`names` は辞書（dict）で、`:` の左の銘柄コードに右の名前を対応させています。`names["01110"]` は `"Auto A"` です。`map(names)` は、`Code` の値ごとに辞書を引いて名前に置き換えます。

やってみる

業種の辞書 `{"01110": "Auto", "02220": "Auto", "03330": "Retail"}` を作り、`prices["Sector"]` に入れて `prices["Sector"].value_counts()` を実行します。Autoが976行、Retailが488行です。

In [42]:
# ここに書きます

## 5-6　確認表を作る

読み方と絞り込みによって、行数・型・欠損がどう変わったかを1つの表に残します。

実行する

In [43]:
raw = pd.read_csv("data/prices.csv")

check = pd.DataFrame({
    "step": ["read only", "dtype and parse_dates", "2024-08 only"],
    "rows": [len(raw), len(prices), len(aug)],
    "Code dtype": [raw["Code"].dtype, prices["Code"].dtype, aug["Code"].dtype],
    "Date dtype": [raw["Date"].dtype, prices["Date"].dtype, aug["Date"].dtype],
    "missing C": [raw["C"].isna().sum(), prices["C"].isna().sum(), aug["C"].isna().sum()],
    "Vo == 0": [(raw["Vo"] == 0).sum(), (prices["Vo"] == 0).sum(), (aug["Vo"] == 0).sum()],
})
check

,step,rows,Code dtype,Date dtype,missing C,Vo == 0
0,read only,1545,int64,object,1,1
1,dtype and parse_dates,1545,object,datetime64[ns],1,1
2,2024-08 only,66,object,datetime64[ns],1,1


何も付けずに読むと `Code` は `int64`、`Date` は `object` です。`C` の欠損と出来高0はどの段階でも1つずつで、Retail Cの2024年8月5日です。8月だけに絞っても、この日は含まれています。

やってみる

5-3のやってみるで作った `y2025` で、3行目を置き換えます。`"2024-08 only"` を `"2025 only"` に、`aug` を `y2025` に変えると、行数771、欠損0、出来高0の行も0になります。

In [44]:
# ここに書きます

# 6. 図を描く

## 6-1　1社の終値

実行する

In [45]:
import matplotlib.pyplot as plt

auto_a = prices[prices["Code"] == "01110"]

fig, ax = plt.subplots(figsize=(8, 4))
ax.plot(auto_a["Date"], auto_a["C"])
ax.set_title("Auto A: closing price, 2024-01-04 to 2025-12-30")
ax.set_xlabel("Date")
ax.set_ylabel("Yen per share")
fig.text(0.99, 0.01, "Source: Fictional data", ha="right", fontsize=8)
plt.show()

Matplotlib is building the font cache; this may take a moment.


/var/folders/7q/9m3zdcp550d685wgsq9kvf6r0000gn/T/ipykernel_87381/3191419479.py:11: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


`plt.subplots` で図の枠（`fig`）と座標軸（`ax`）を作り、`ax.plot(横軸, 縦軸)` で折れ線を描きます。タイトルに銘柄と期間、縦軸に単位、右下に出典を書いておくと、図だけを見ても何の数値かが分かります。ラベルが英語なのは、日本語のfontがないPCで文字が四角になるのを避けるためです。

やってみる

1. 下のcellにこのcodeをコピーし、`"01110"` を `"03330"` に、タイトルの `Auto A` を `Retail C` に変えます。
2. さらに `ax.plot` の `"C"` を `"Vo"` に、縦軸のラベルを `"Shares traded"` に変えます。2024年8月5日に、出来高が0まで下がる線が出ます。

In [46]:
# ここに書きます

## 6-2　株式分割と調整後終値

Auto B（`02220`）は、2025年4月1日に1株を2株に分割しました。

実行する

In [47]:
split = prices[(prices["Code"] == "02220") & (prices["Date"] >= "2025-03-27") & (prices["Date"] <= "2025-04-03")]
split[["Date", "C", "AdjFactor", "AdjC", "Vo", "AdjVo"]]

,Date,C,AdjFactor,AdjC,Vo,AdjVo
831,2025-03-27,2217.0,1.0,1108.5,9647600.0,19295200.0
832,2025-03-28,2239.0,1.0,1119.5,12979300.0,25958600.0
833,2025-03-31,2227.0,1.0,1113.5,9403300.0,18806600.0
834,2025-04-01,1117.0,0.5,1117.0,26871400.0,26871400.0
835,2025-04-02,1112.0,1.0,1112.0,22040800.0,22040800.0
836,2025-04-03,1133.0,1.0,1133.0,14814700.0,14814700.0


3月31日の終値2,227円が、4月1日には1,117円になっています。1株が2株になったので、1株の価格が約半分になりました。`AdjFactor` はこの日だけ0.5です。`AdjC` は分割前の価格に0.5を掛けた値で、3月31日は1,113.5円です。出来高も、分割前の株数を2倍にした `AdjVo` があります。

実行する

In [48]:
auto_b = prices[(prices["Code"] == "02220") & (prices["Date"] >= "2025-01-01") & (prices["Date"] <= "2025-06-30")]

fig, ax = plt.subplots(figsize=(8, 4))
ax.plot(auto_b["Date"], auto_b["C"], label="C (unadjusted)")
ax.plot(auto_b["Date"], auto_b["AdjC"], label="AdjC (adjusted for the split)")
ax.axvline(pd.Timestamp("2025-04-01"), color="gray", linestyle="--")
ax.set_title("Auto B: 2-for-1 split on 2025-04-01, 2025-01 to 2025-06")
ax.set_xlabel("Date")
ax.set_ylabel("Yen per share")
ax.legend()
fig.text(0.99, 0.01, "Source: Fictional data", ha="right", fontsize=8)
plt.show()

/var/folders/7q/9m3zdcp550d685wgsq9kvf6r0000gn/T/ipykernel_87381/3027772185.py:12: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


`C` は4月1日に段差ができ、`AdjC` はつながっています。`label` に書いた名前が、`ax.legend()` で凡例に出ます。灰色の破線が分割の日です。

やってみる

`auto_b` の期間を2024年1月4日から2025年12月30日に広げます。`C` だけを見ると、Auto Bの株価が2年で3,000円から1,440円へ半分以下になったように見えます。`AdjC` では1,500円から1,440円です。

In [49]:
# ここに書きます

## 6-3　3社をforで重ねる

実行する

In [50]:
fig, ax = plt.subplots(figsize=(8, 4))
for code in ["01110", "02220", "03330"]:
    one = prices[prices["Code"] == code]
    ax.plot(one["Date"], one["AdjC"], label=names[code])
ax.set_title("Adjusted close of three firms, 2024-01-04 to 2025-12-30")
ax.set_xlabel("Date")
ax.set_ylabel("Yen per share")
ax.legend()
fig.text(0.99, 0.01, "Source: Fictional data", ha="right", fontsize=8)
plt.show()

/var/folders/7q/9m3zdcp550d685wgsq9kvf6r0000gn/T/ipykernel_87381/3975108246.py:10: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


`for` の1回ごとに1社を選び、線を1本描いています。3社は株価の水準が違うので、値動きの大きさを比べにくい図です。

やってみる

listから `"02220"` を消して、2社だけの図にします。

In [51]:
# ここに書きます

## 6-4　最初の日を100にそろえる

実行する

In [52]:
fig, ax = plt.subplots(figsize=(8, 4))
for code in ["01110", "02220", "03330"]:
    one = prices[prices["Code"] == code]
    base = one["AdjC"].iloc[0]
    ax.plot(one["Date"], one["AdjC"] / base * 100, label=names[code])
ax.axhline(100, color="gray", linewidth=0.8)
ax.set_title("Adjusted close, 2024-01-04 = 100")
ax.set_xlabel("Date")
ax.set_ylabel("Index (2024-01-04 = 100)")
ax.legend()
fig.text(0.99, 0.01, "Source: Fictional data", ha="right", fontsize=8)
plt.show()

/var/folders/7q/9m3zdcp550d685wgsq9kvf6r0000gn/T/ipykernel_87381/587042281.py:12: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


`iloc[0]` は、その表の最初の行の値です。各社の調整後終値を最初の日の値で割って100を掛けると、3社とも100から始まり、値動きの割合を比べられます。

やってみる

`for` の中に `print(code, round(one["AdjC"].iloc[-1] / base * 100, 1))` を足します。`iloc[-1]` は最後の行です。`01110` が113.8、`02220` が96.0、`03330` が113.8と表示されます。

In [53]:
# ここに書きます

## 6-5　日次の変化率のヒストグラム

2024年度の教材と同じく、前の日からの変化率の分布を見ます。

実行する

In [54]:
auto_a = prices[prices["Code"] == "01110"].copy()
auto_a["change"] = (auto_a["AdjC"] - auto_a["AdjC"].shift(1)) / auto_a["AdjC"].shift(1) * 100
print(auto_a["change"].describe())

fig, ax = plt.subplots(figsize=(8, 4))
ax.hist(auto_a["change"].dropna(), bins=40)
ax.set_title("Auto A: daily change of adjusted close, 2024-01-05 to 2025-12-30")
ax.set_xlabel("Daily change (%)")
ax.set_ylabel("Number of days")
fig.text(0.99, 0.01, "Source: Fictional data", ha="right", fontsize=8)
plt.show()

count    514.000000
mean       0.037818
std        1.596782
min       -4.526889
25%       -0.965507
50%        0.071327
75%        1.061691
max        5.146533
Name: change, dtype: float64


/var/folders/7q/9m3zdcp550d685wgsq9kvf6r0000gn/T/ipykernel_87381/3760627183.py:11: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


`shift(1)` は1行上の値、つまり前の営業日の値です。1-2の `pct_change(old, new)` と同じ計算を、列全体に一度に行っています。最初の日は前の日がないので `NaN` になり、`count` は487です。平均は約0.04%、最小は約-4.85%、最大は約5.15%です。

`.copy()` を付けると、`prices` とは別の表として `auto_a` に列を足せます。

やってみる

`"01110"` を `"02220"` に変え、`change` の行にある3つの `"AdjC"` を `"C"` に変えます。-49.84%の日が1日だけ、他から離れて現れます。2025年4月1日の分割の日です。`"AdjC"` のままなら、この日は0.31%です。

In [55]:
# ここに書きます

## 6-6　図の1点をCSVの行まで戻る

実行する

In [56]:
auto_a = prices[prices["Code"] == "01110"]
i = auto_a["AdjC"].idxmax()
print(i)
print(prices.loc[i, ["Date", "Code", "AdjC"]])

510
Date    2025-12-24 00:00:00
Code                  01110
AdjC                 3119.0
Name: 510, dtype: object


`idxmax()` は、最大値がある行の番号（index）を返します。Auto Aの調整後終値の最大は、2025年12月24日の3,119.0円で、行番号は483です。

実行する

In [57]:
fig, ax = plt.subplots(figsize=(8, 4))
ax.plot(auto_a["Date"], auto_a["AdjC"])
ax.plot(prices.loc[i, "Date"], prices.loc[i, "AdjC"], "o", color="red")
ax.set_title("Auto A: adjusted close and its maximum, 2024-01-04 to 2025-12-30")
ax.set_xlabel("Date")
ax.set_ylabel("Yen per share")
fig.text(0.99, 0.01, "Source: Fictional data", ha="right", fontsize=8)
plt.show()

/var/folders/7q/9m3zdcp550d685wgsq9kvf6r0000gn/T/ipykernel_87381/3332047822.py:8: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


赤い点が最大の日です。この点をCSV fileの行まで戻ります。

実行する

In [58]:
lines = open("data/prices.csv", encoding="utf-8").read().splitlines()
print(lines[0])
print(lines[i + 1])

Date,Code,O,H,L,C,Vo,Va,AdjFactor,AdjO,AdjH,AdjL,AdjC,AdjVo
2025-12-24,01110,3083.0,3121.0,3081.0,3119.0,32219400.0,100105676000.0,1.0,3083.0,3121.0,3081.0,3119.0,32219400.0


`prices` は `data/prices.csv` を並べ替えずに読んだので、行番号0がCSVの2行目（見出しの次）にあたります。行番号483はCSVの485行目で、0から数えるlistの `lines` では `lines[484]` です。表示された行の `AdjC`（右から2つ目）は3119.0で、図の赤い点と同じです。

VS Codeで `data/prices.csv` を開き、⌃G（WindowsはCtrl+G）で485と入れると、同じ行に移動します。

やってみる

Auto B（`"02220"`）で、`AdjC` の最大と `C` の最大を探します。`AdjC` の最大は2025年11月27日の1,689.0円で、CSVの954行目です。`C` の最大は2024年4月5日の3,210.0円で、分割前の価格です。

In [59]:
# ここに書きます

# 7. 関数にまとめる

## 7-1　読み込みの関数

実行する

In [60]:
def load_prices(path):
    frame = pd.read_csv(path, dtype={"Code": str}, parse_dates=["Date"])
    frame = frame.sort_values(["Code", "Date"]).reset_index(drop=True)
    return frame

prices2 = load_prices("data/prices.csv")
print(prices2.shape)
print(prices2.dtypes)

(1545, 14)
Date         datetime64[ns]
Code                 object
O                   float64
H                   float64
L                   float64
C                   float64
Vo                  float64
Va                  float64
AdjFactor           float64
AdjO                float64
AdjH                float64
AdjL                float64
AdjC                float64
AdjVo               float64
dtype: object


4節と5節の読み方（`dtype`、`parse_dates`、並べ替え）を1つの関数にしました。同じ列を持つ別のCSVでも、pathを変えるだけで同じ読み方になります。

やってみる

`load_prices("data/prices_short.csv")` を実行します。`Code` は文字列、`Date` は日付型で読めますが、`C` は `object` のままです。`,` を消す処理が関数に入っていないからです。

In [61]:
# ここに書きます

## 7-2　銘柄と期間を引数にした図の関数

実行する

In [62]:
def plot_normalized(df, codes, start, end, source="Fictional data"):
    period = df[df["Date"].between(start, end)].sort_values("Date")
    fig, ax = plt.subplots(figsize=(8, 4))
    plotted = 0
    for code in codes:
        one = period[period["Code"] == code].dropna(subset=["AdjC"])
        if one.empty or one["AdjC"].iloc[0] <= 0:
            print(f"{code}: 正の基準値がありません。銘柄・期間・欠損を確認します。")
            continue
        base = one["AdjC"].iloc[0]
        base_date = one["Date"].iloc[0].date()
        ax.plot(one["Date"], one["AdjC"] / base * 100, label=f"{code} (base {base_date})")
        plotted += 1
    if not plotted:
        ax.text(0.5, 0.5, "No valid observations", transform=ax.transAxes, ha="center")
    ax.axhline(100, color="gray", linewidth=0.8)
    ax.set_title("Adjusted close, first trading day = 100, " + start + " to " + end)
    ax.set_xlabel("Date")
    ax.set_ylabel("Index (first trading day = 100)")
    ax.legend()
    fig.text(0.99, 0.01, "Source: " + source, ha="right", fontsize=8)
    return ax

plot_normalized(prices2, ["01110", "02220", "03330"], "2024-01-04", "2025-12-30")
plt.show()

/var/folders/7q/9m3zdcp550d685wgsq9kvf6r0000gn/T/ipykernel_87381/926557855.py:25: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


6-4と同じ図です。`df`、`codes`、`start`、`end` を変えると、別の表・銘柄・期間で同じ形の図ができます。`source="Fictional data"` のように `=` で値を書いた引数は、呼ぶときに省略するとその値が使われます。`dropna()` で欠損を除いてから最初の値を取るので、期間の初日の価格が欠損なら、次の日の価格が100になります。

使う列は `Date`、`Code`、`AdjC` だけです。第8回でJ-Quantsから取得する表にも同じ列があるので、この関数をそのまま使えます。

やってみる

1. `plot_normalized(prices2, ["01110", "03330"], "2025-01-01", "2025-06-30")` を実行します。2社・半年の図になり、6月30日はAuto Aが79.9、Retail Cが87.5です。
2. `codes` を `["03330"]`、`start` を `"2024-08-05"`、`end` を `"2024-08-30"` にします。8月5日は価格が欠損なので、線は8月6日から始まり、8月6日が100です。

In [63]:
# ここに書きます

# 8. 最初から実行し直す

1. notebook上部の Restart を押します。kernelの変数がすべて消えます。
2. Run All を押します。上のcellから順に、全部が実行されます。
3. errorで止まったら、そのcellでerrorを確かめた行に `#` が付いているかを見ます。

4-6の `output/prices_short_clean.csv` は、同じ内容で上書きされます。CSVを読む、型を直す、並べ替える、図を描く、の順に上から実行され、同じ表と図がもう一度できます。